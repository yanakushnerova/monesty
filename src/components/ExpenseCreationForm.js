import React from "react"
import moment from "moment"
import { SingleDatePicker } from "react-dates"
import 'react-dates/initialize'

export default class ExpenseCreationForm extends React.Component {
    constructor (props) {
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }

    onAmountChange = (e) => {
        const amount = e.target.value

        if (amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }

    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }

    onSubmit = (e) => {
        e.preventDefault()

        if (!this.state.description || ! this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf()
            })
        }
    }

    render() {
        return (
            <div>
                { this.state.error && <p>{this.state.error}</p> }
                <form onSubmit={this.onSubmit}>
                    <input 
                        type='text'
                        placeholder='description' 
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        autoFocus 
                    />
                    <input
                        type='text'
                        placeholder='amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea 
                        placeholder='Add a note (optional)'
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    
                    { this.props.expense ? <button>edit expense</button> : <button>add expense</button> }
                </form>
            </div>
        )
    }
}
