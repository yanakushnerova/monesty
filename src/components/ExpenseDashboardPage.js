import React from 'react'
import ExpenseList from './ExpenseList'

const ExpenseDashboardPage = () => (
	<div>
		<div>this is my ExpenseDashboardPage</div>
		<ExpenseList />
	</div>
)

export default ExpenseDashboardPage
