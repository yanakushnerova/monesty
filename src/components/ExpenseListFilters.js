import React from "react"
import { connect } from "react-redux"
import { setTextFilter, sortByAmount, sortByDate, setEndDate, setStartDate } from "../actions/filters"
import { DateRangePicker } from 'react-dates'
import 'react-dates/initialize'

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate()
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount()
        }
    }

    render () {
        return (
            <div>
                <input type='text' defaultValue={this.props.filters.text} onChange={this.onTextChange}></input>

                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value='date'>date</option>
                    <option value='amount'>amount</option>
                </select>

                <DateRangePicker
                    className="dateRangePicker"
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    startDateId="start"
                    endDateId="end"
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date)),
    dispatch
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
