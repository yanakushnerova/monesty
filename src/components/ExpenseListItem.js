import React from 'react'
import { useHistory } from 'react-router-dom'

const ExpenseListItem = ({id, description, amount, createdAt}) => {
    const history = useHistory()
    
    return (
            <div>
                <p>{description}</p>
                <p>{amount} - {createdAt}</p>
                <button onClick={() => {
                     history.push(`/edit/${id}`)
                }}>edit expense</button>
            </div>
    )
}

export default ExpenseListItem
