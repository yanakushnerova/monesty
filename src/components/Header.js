import React from "react"
import { NavLink } from "react-router-dom"

const Header = () => (
	<div>
		<h1>title</h1>
		<NavLink to='/' activeClassName="is-active">dashboard</NavLink>
		<NavLink to='/create' activeClassName="is-active">create expense</NavLink>
		<NavLink to='/edit' activeClassName="is-active">edit expense</NavLink>
		<NavLink to='/help' activeClassName="is-active">help</NavLink>
	</div>
)

export default Header
