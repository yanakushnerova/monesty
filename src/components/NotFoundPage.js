import React from "react"
import { Link } from "react-router-dom"

const NotFoundPage = () => (
	<div>not found, go <Link to='/'>home</Link></div>
)

export default NotFoundPage
