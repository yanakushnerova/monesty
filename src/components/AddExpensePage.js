import React from 'react'
import { connect } from 'react-redux'
import ExpenseCreationForm from './ExpenseCreationForm'
import { addExpense } from '../actions/expenses'

export const AddExpensePage = (props) => (
	<div>
		AddExpensePage
		<ExpenseCreationForm 
			onSubmit={(expense) => {
				props.onSubmit(expense)
				props.history.push('/')
			}}
		/>
	</div>
)

const mapDispatchToProps = (dispatch) => ({
	onSubmit: (expense) => {
		dispatch(addExpense(expense))
}})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
