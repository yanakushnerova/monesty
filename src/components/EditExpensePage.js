import React from 'react'
import { connect } from 'react-redux'
import ExpenseCreationForm from './ExpenseCreationForm'
import { editExpense, removeExpense } from '../actions/expenses'

export const EditExpensePage = (props) => {
	const onSubmit = (expense) => {
		props.onSubmit(props.expense.id, expense)
		props.history.push('/')
	}

	const onRemove = () => {
		props.onRemove({ id: props.expense.id })
		props.history.push('/')
	}

	return (
		<div>
			<ExpenseCreationForm expense={props.expense} onSubmit={onSubmit} />
			<button onClick={onRemove}>remove</button>
		</div>
	)
}

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find((expense) => {
		return expense.id === props.match.params.id
	})
})

const mapDispatchToProps = (dispatch, props) => ({
	onSubmit: (id, expense) => dispatch(editExpense(id, expense)),
	onRemove : (data) => dispatch(removeExpense(data)),
	dispatch
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
