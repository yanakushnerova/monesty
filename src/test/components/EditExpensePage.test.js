import React from "react"
import { shallow } from "enzyme"
import { EditExpensePage } from "../../components/EditExpensePage"
import expenses from '../fixtures/expenses'

let onSubmit, onRemove, history, wrapper

beforeEach(() => {
    onSubmit = jest.fn()
    onRemove = jest.fn()
    history  = { push: jest.fn() }
    wrapper = shallow(<EditExpensePage onSubmit={onSubmit} onRemove={onRemove} history={history} expense={expenses[2]} />)
})

test('should render edit expense page', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should handle edit expense', () => {
    wrapper.find('ExpenseCreationForm').prop('onSubmit')(expenses[2])
    expect(onSubmit).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])
    expect(history.push).toHaveBeenLastCalledWith('/')
})

test('should handle remove expense', () => {
    wrapper.find('button').simulate('click')
    expect(onRemove).toHaveBeenLastCalledWith({ id: expenses[2].id })
    expect(history.push).toHaveBeenLastCalledWith('/')
})
