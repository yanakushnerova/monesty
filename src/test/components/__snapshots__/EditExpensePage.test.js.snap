// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page 1`] = `
<div>
  <ExpenseCreationForm
    expense={
      Object {
        "amount": 2400,
        "createdAt": 100000000,
        "description": "present",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
